### PR TITLE
[BWS][REF]: Push notifications code refactor

### PR DESCRIPTION
**Pull request by rastajpa**

### CODE REVIEW

File: packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts [hunk 1]
```diff
@@ -498,8 +505,9 @@ export class PushNotificationsService {
 
           this.storage.fetchPushNotificationSubs(recipient.copayerId, (err, subs) => {
             if (err) return next(err);
-            if(subs && subs[0]) subs[0].plain = content.plain;
-            return next(err, subs);
+
+            const subscriptions = subs.length ? subs.map(obj => ({ ...obj, plain: content.plain })) : subs;
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts:509]: will fail is subs = null

**Comment by SatoshiNakamotogmx** [comment]
Merge is Blocked

**Review by matiu — APPROVED** [review]

File: packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts [hunk 2]
```diff
@@ -37,8 +37,26 @@ const PUSHNOTIFICATIONS_TYPES = {
     filename: 'txp_finally_rejected'
   },
   TxConfirmation: {
-    filename: 'tx_confirmation',
-    notifyCreatorOnly: true
+    filename: 'tx_confirmation'
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts:40]: notifyCreatorOnly should stay.

File: packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts [hunk 3]
```diff
@@ -443,6 +467,38 @@ export class PushNotificationsService {
     };
   }
 
+  _getSubscriptions(notification, notifType, recipientsList, contents, cb) {
+    if (!notifType.broadcastToActiveUsers) {
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts:471]: do the true loop first.

File: packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts (hunk 2 above)
Comment (rastajpa) [reply packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts:40]: It is no longer necessary to keep that variable as notifications must also be sent to the creator. That variable was only needed in this part of the code that was removed:
 if (
        (copayer.id == notification.creatorId && notificationType.notifyCreatorOnly) ||
        (copayer.id != notification.creatorId && !notificationType.notifyCreatorOnly)
   )

File: packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts [hunk 4]
```diff
@@ -153,51 +171,56 @@ export class PushNotificationsService {
...
+                const notificationData: any = {
+                  to: sub.token,
+                  priority: 'high',
+                  restricted_package_name: sub.packageName,
+                  data: {
+                    walletId: sjcl.codec.hex.fromBits(sjcl.hash.sha256.hash(notification.walletId || sub.walletId)),
+                    tokenAddress,
+                    multisigContractAddress,
+                    copayerId: sjcl.codec.hex.fromBits(sjcl.hash.sha256.hash(sub.copayerId)),
+                    title: sub?.plain?.subject,
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts:194]: maybe is more clear to add `title` and `body` members in the if bellow

File: packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts [hunk 5]
```diff
@@ -153,51 +171,56 @@ export class PushNotificationsService {
...
+                    body: sub?.plain?.body,
+                    notification_type: notification.type
+                  }
+                };
+
+                if (!notifType.dataOnly) {
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts:200]: here.
notification.title = ...
notification.body = ...

File: packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts [hunk 6]
```diff
@@ -443,6 +467,38 @@ export class PushNotificationsService {
...
   }
 
+  _getSubscriptions(notification, notifType, recipientsList, contents, cb) {
+    if (notifType.broadcastToActiveUsers) {
+      this.storage.fetchLatestPushNotificationSubs((err, subs) => {
+        if (err) return cb(err);
+
+        const allSubs = _.reject(subs, sub => !sub.walletId);
+        logger.info(
+          `Sending NewBlock [${notification.data.coin}/${notification.data.network}] notifications to: ${allSubs.length} subscribers`
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts:477]: change `NewBlock` to `${notification.type}`

File: packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts [hunk 7]
```diff
@@ -443,6 +467,38 @@ export class PushNotificationsService {
     };
   }
 
+  _getSubscriptions(notification, notifType, recipientsList, contents, cb) {
+    if (notifType.broadcastToActiveUsers) {
+      this.storage.fetchLatestPushNotificationSubs((err, subs) => {
+        if (err) return cb(err);
+
+        const allSubs = _.reject(subs, sub => !sub.walletId);
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts:475]: nice.

File: packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts [hunk 8]
```diff
@@ -259,39 +284,38 @@ export class PushNotificationsService {
...
-              return {
-                copayerId: copayer.id,
-                language: p.language || this.defaultLanguage,
-                unit: unit || p.unit || this.defaultUnit
-              };
-            }
-          })
-        );
-
+        const recipientsList = wallet
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts:295]: intead of this `wallet` check and this https://github.com/bitpay/bitcore/pull/3027/files#diff-516ac13e45f6e22ca8d66a9b72e8d20851abddac3e12637d389862a8b95295e5R270


maybe add:

`if (!wallet) return cb(null, []); `

at: https://github.com/bitpay/bitcore/pull/3027/files#diff-516ac13e45f6e22ca8d66a9b72e8d20851abddac3e12637d389862a8b95295e5R268

File: packages/bitcore-wallet-service/test/integration/server.js [hunk 9]
```diff
@@ -8933,6 +8933,7 @@ describe('Wallet service', function() {
           token: 'DEVICE_TOKEN',
           packageName: 'com.wallet',
           platform: 'Android',
+          walletId: '123'
```
Comment (matiu) [inline packages/bitcore-wallet-service/test/integration/server.js:8936]: keep the old tests (As the old clients should still work) and add a new tests with the new `walletId` member.

**Pull request merged**